core/src/main/java/io/questdb/cairo/wal/ApplyWal2TableJob.java

```diff
@@ -486,91 +493,83 @@ private void storeWalTelemetryNoop(short event, TableToken tableToken, int walId
+                    }
+                    // Do not suspend table. Perhaps writer will be unlocked with no transaction applied.
+                    // We do not suspend table because of having initial value on lastWriterTxn. It will either be
+                    // "ignore" or last txn we applied.
+                    return;
                 }
-                // Don't suspend table. Perhaps writer will be unlocked with no transaction applied.
-                // We don't suspend table by virtue of having initial value on lastWriterTxn. It will either be
-                // "ignore" or last txn we applied.
-                return lastWriterTxn;
-            }
 
-            if (engine.getTableSequencerAPI().notifyCommitReadable(tableToken, lastWriterTxn)) {
-                engine.notifyWalTxnCommitted(tableToken);
+                if (engine.getTableSequencerAPI().notifyCommitReadable(tableToken, lastWriterTxn)) {
+                    engine.notifyWalTxnCommitted(tableToken);
+                }
             }
         } catch (CairoException ex) {
             if (ex.isTableDropped() || engine.isTableDropped(tableToken)) {
                 engine.notifyDropped(tableToken);
                 // Table is dropped, and we received cairo exception in the middle of apply
-                return tryDestroyDroppedTable(tableToken, null, engine, tempPath) ? Long.MAX_VALUE : WAL_APPLY_IGNORE_ERROR;
+                tryDestroyDroppedTable(tableToken, null, engine, tempPath);
+            } else {
+                telemetryFacade.store(TelemetrySystemEvent.WAL_APPLY_SUSPEND, TelemetryOrigin.WAL_APPLY);
+                LOG.critical().$("job failed, table suspended [table=").utf8(tableToken.getDirName())
+                        .$(", error=").$(ex.getFlyweightMessage())
+                        .$(", errno=").$(ex.getErrno())
+                        .I$();
+                handleWalApplyFailure(tableToken, ex.isOutOfMemory()
+                        ? new WalError(ex.getErrno(), Tag.FAILED_MEMORY_ALLOCATION, ex.getFlyweightMessage())
+                        : new WalError(ex.getErrno(), ex.getFlyweightMessage()));
             }
-            telemetryFacade.store(TelemetrySystemEvent.WAL_APPLY_SUSPEND, TelemetryOrigin.WAL_APPLY);
-            LOG.critical().$("job failed, table suspended [table=").utf8(tableToken.getDirName())
-                    .$(", error=").$(ex.getFlyweightMessage())
-                    .$(", errno=").$(ex.getErrno())
-                    .I$();
-            return WAL_APPLY_FAILED;
         } catch (Throwable ex) {
             telemetryFacade.store(TelemetrySystemEvent.WAL_APPLY_SUSPEND, TelemetryOrigin.WAL_APPLY);
             LOG.critical().$("job failed, table suspended [table=").utf8(tableToken.getDirName())
                     .$(", error=").$(ex)
                     .I$();
-            return WAL_APPLY_FAILED;
+            handleWalApplyFailure(tableToken, new WalError(-1, Tag.OTHER, ex.getMessage()));
```

Comment:
lets not create extra objects please:
- we need to align types in the tracker and parameters, message should be `String` explicitly to avoid passing mutable `CharSequence` objects. I will point at a bug that was caused by this misalignment already.
- exception message always includes `errno`, i think storing errno in separate field is not necessary (it is there twice effectively)
- I would remove `WalError` as an object, the methods can be static
-

Reply:
originally i had separate fields, but then decided to wrap the fields into an object to avoid threading issues.

in `SeqTxnTracker` we need to make sure that the fields are becoming visible to other threads at the same time, so added a single volatile `walError`:
```
private volatile WalError walError;
```
instead of:
```
private volatile int errorCode;
private volatile WalErrorTag errorTag;
private volatile CharSequence errorMessage;
```

even after removing the error code, we still have two fields to set.

i do not think we would create too many WalError objects, as it will be created only when a table is suspended.
should i keep `WalError`?
i think it is the simplest way to avoid visibility issues.

---------

holding onto the `CharSequence` object from CairoException is definitely a bad idea.
thanks for pointing it out, fixing it.
will store the error message as a string as you suggested.

Reply:
i have an idea now how to do this without the `WalError` object.
i will set tag and error message first, then flip the suspended state which is also volatile.
tag and error message will be read in `WalTableListFunctionFactory` only if the table is suspended, so will not read the fields before everything is set.
will add comments to make sure the order of setting the tag, the error message and the suspended state will not be changed accidentally.

Reply:
done